### ISSUE TITLE
Question - Cross communication between Reaper and Reapy

### ISSUE THREAD

**Issue by DrewTNBD** [RomeoDespres/reapy#86]
I'm writing some wrapper classes around the Reapy classes so I can pull all of the best bits of Numpy, SciPy, and Librosa into a selection of "media item analysis classes". It's working quite well when running PycharmCE and Reapy. I can do all the stuff I want to do.

Now I'm thinking about how to interface with Reaper, and trigger methods within these classes from Reaper.

Is it possible to have any kind of cross communication between Reaper and Reapy?

Take this small example:
```
import reapy as rp

class ReapyConnection:

    def __init__(self):
        self.project = rp.Project()

    def get_Selected_Items(self):
        with rp.inside_reaper():
            selected_items = self.project.selected_items
            rp.show_console_message(selected_items)

while True:
    con = ReapyConnection()
```

I have an object called con, which is instantiating an instance of the ReapyConnection class. What I want to do is from inside Reaper tell that instance to call it's get_Selected_Items() method.

Is this possible?

**Comment by Levitanus** [comment]
What do you mean by «cross communication»? You can poll reaper wherever you want from wherever you want: from reaper itself, from localhost, from Android tablet etc, this doesn't need any extra stuff. 
Only thing you need for something other than localhost is `reapy.connect('ip')`. Or I didn't get you desire?

**Comment by DrewTNBD** [comment]
Sorry I probably was not clear.

I want to address Reapy that is running outside Reaper, but from within Reaper. For the purposes of feeding data from Reaper to an external Python process - in my case a Kivy app.

It sounds like I want to investigate this reapy.connect('ip') method that you've mentioned.

**Comment by dimentorium** [comment]
Hi DrewTNBD,
sounds like the same I do, so having buttons in Reaper that control my Python/TKInter Program. In my case, I run a TCP server in my app, and from Reaper I send strings to trigger actions. Maybe tht helps you. Please check if Kivy has some specialties on threading; as the server runs in the background Qt will not accept calls and needs some special code, TKinter does not care about

@Levitanus: is that also directly possible in Reapy?
@DrewTNBD : if  you are interested in exchanging, I am working on a Presetmanager for Reaper where I want to embed automated tagging and Classification of Synthesizer sounds with python libraries. 

===============Server====================
import socket
import threading
import logging
from core import actions

rs = None

def start():
    global rs
    rs = reaper_server()


class reaper_server():
    """reaper_server.

    TCP/IP server waiting for commands from Reaper

    Methods
    -------
        init: init class
        run: start server and wait for messages

    Properties
    ----------
        host: host address
        port: TCP port to listen
    """

    def __init__(self):
        """Init.

        Set TCP settings in class
        """
        logging.debug("Server started")
        # Standard loopback interface address (localhost)
        self.__host = '127.0.0.1' 
        # Port to listen on (non-privileged ports are > 1023) 
        self.__port = 65432 

        thread = threading.Thread(target=self.run, args=())
        thread.daemon = True
        thread.start()

    def run(self):
        """Run.

        Run server
        """
        logging.debug("Starting Reaper Com Server")
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.bind((self.__host, self.__port))
            while True:
                logging.debug("Waiting")
                s.listen()
                conn, addr = s.accept()
                with conn:
                    self.__connected = True
                    logging.debug('Connected by' + str(addr))
                    while self.__connected:
                        data = conn.recv(1024)
                        logging.debug("Server" + str(data))
                        if not data:
                            logging.debug('Disconnected')
                            self.__connected = False
                        message = data.decode("utf-8")
                        #implement function calls
                        if message == "save_preset":
                            actions.save_preset()
                        elif message == "show":
                            actions.show()

=======Reaper============
#!/usr/bin/env python3
import re
import socket

HOST = '127.0.0.1'  # The server's hostname or IP address
PORT = 65432        # The port used by the server

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.connect((HOST, PORT))
    s.sendall(b'show')

**Comment by RomeoDespres** [comment]
The `connect(ip)` will only help if you want to run a reapy script outside REAPER on another device than the one that runs REAPER. For instance, an Android app written in Python could use reapy to control a REAPER instance running on a laptop of your local network.

I don't think this is your question though. It has previously been asked on the forum, here is what I replied there:

> Hmm I guess it should be somehow possible. You can add any action you want to a toolbar by customizing it, so you can either:
>
> 1. run your reapy script from inside REAPER directly with that action,
> 2. or, if you need your reapy app to live outside REAPER, use the toolbar action to simply "send a message" to reapy (for instance by using reapy.set_ext_state, or with your own logic that could include files, sockets, flask, etc.)

If you want more details about how to implement these solutions feel free to ask.

**Comment by RomeoDespres** [comment]
Thanks @dimentorium, you wrote it faster than I!

Just for clarification in regard to what @Levitanus said: the `connect(ip)` will only help if you want to run a reapy script outside REAPER on another device than the one that runs REAPER. For instance, an Android app written in Python could use reapy to control a REAPER instance running on a laptop of your local network. I believe this is a different issue than yours.

**Comment by RomeoDespres** [comment]
I would also suggest looking at simple HTTP servers like `flask` or `tornado` if you don't want to dive into sockets too much, because these are very low level and can be painful. HTTP servers make communicating over the network quite simpler. And since the server has to be on the outside in your case (and not inside REAPER), it doesn't imply using third-party libraries inside REAPER (which makes it crash).

**Comment by drewfxpansion** [comment]
Nice one guys. Yes it sounds like `connect(ip)` is not what is needed here. Basically I'm envisioning either a small Kivy app with a bunch of buttons that can trigger various functions within Reaper, using Python libraries running in the Kivy app.

@dimentorium I read about your project on the Reaper forums (I'm Amberience over there) and it sounds very cool. I'm not sure how familiar with MIR you are, but for classification based on audio feature extraction, that'd be the way I would go. Get a bunch of synth presets and pull out spectral centroid, rolloff, contrast, etc and use those features for some sort of clustering.

So for the XLNAudio XO style "feature map" that's probably the direction I would head in.

**Comment by DrewTNBD** [comment]
Nice one guys. Yes it sounds like connect(ip) is not what is needed here. Basically I'm envisioning either a small Kivy app with a bunch of buttons that can trigger various functions within Reaper, using Python libraries running in the Kivy app. I will definitely look into TCP/IP server, and maybe flask could do it.

Funnily enough, all this work is precisely because Reaper crashes out on most 3rd party Python libraries I've tried. I thought it was to do with permissions and/or environment paths, but I addressed all that and it's still flaky. Doing a simply numpy array initialisation just crashes to desktop!

@dimentorium I read about your project on the Reaper forums (I'm Amberience over there) and it sounds very cool. I'm not sure how familiar with MIR you are, but for classification based on audio feature extraction, that'd be the way I would go. Get a bunch of synth presets and pull out spectral centroid, rolloff, contrast, etc and use those features for some sort of clustering.

So for the XLNAudio XO style "feature map" that's probably the direction I would head in.

**Comment by DrewTNBD** [comment]
@RomeoDespres - regarding your post on the Reaper forums:
`or, if you need your reapy app to live outside REAPER, use the toolbar action to simply "send a message" to reapy (for instance by using reapy.set_ext_state, or with your own logic that could include files, sockets, flask, etc.)`

I don't exactly know how the reapy.set_ext_state stuff works to be honest. But I would assume in that situation I'd have a script running in Reaper that did `SetExtState` and that's a one-time run script from a toolbar or key command.

Then in my app I'd be polling for any changes to ext-states by using a timer, which Kivy actually provides functions for. So it'd be something like:

```
    def __init__(self):
        Clock.schedule_interval(self.get_state, 1/20.)

    def get_state(self, frame):
        self.state = reapy.get_ext_state()
```

Something along those lines?

**Comment by Levitanus** [comment]
For the last message: yes, i think so. Since gui has event loop which still does something with reaper, i think it's quite common to put the loop itself inside context of `inside_reaper ()` and poll reaper ext state as it done by your pc for the keyboard, for example.

**Comment by Levitanus** [comment]
By the way, as soon, as we manage some things with my pull requests, there will be awaible method map, making some polls more efficient than with `inside_reaper ()`

**Comment by DrewTNBD** [comment]
Here is what I've got.

Python3 script running inside Reaper:
```
import reapy as rp

rp.set_ext_state('section','key', 'value string', persist=False)
```

Python3 script running in PyCharm:
```
import reapy as rp

while True:
    state = rp.get_ext_state('section', 'key')
    if len(state) > 0:
        print(state)
        rp.delete_ext_state('section', 'key')
    else:
        continue

```

This results in polling the external state and flushing it once we've acknowledged the new value. So I'm fairly sure I can use this logic to trigger methods in my app classes.

@dimentorium I looked at your script, and first I thought I would try just copypasta-ing your script and see how it all worked. But it fell down on the line `from core import actions` - it can't find any module called core.

**Comment by dimentorium** [comment]
Hi Drew,

actions is just my module where I have the functions that are called. Just remove it and replace the  calls in the last If/Else loop.

**Comment by DrewTNBD** [comment]
> Hi Drew,
> 
> actions is just my module where I have the functions that are called. Just remove it and replace the calls in the last If/Else loop.

Ahhh! Understood! I have a similar one, so will replace.

**Comment by dimentorium** [comment]
What about the idea building a basic framework for a Reapy based toolbox? 
So something with UI, event handling and communication to Reaper to make development of addons and plugins easier? From what I read it looks like we are building comparable stuff

**Comment by Levitanus** [comment]
I'm still dreaming of dockable gui. Even dived into win32, juliansanders api and libx library, but cannot make them work...
Technically, there should not be difference where the script executed from, but, at least at Linux, my attempts to make reaper see my windows or see reaper windows from python fails...

**Comment by dimentorium** [comment]
Hmmm in my old days I used the Windows handle in Java to dock LabVIEW UIs inside it; this also worked in C# and C++. Is it possible to somehow get the Window handle from a Reaper Lua Window? Or create an empty Window with a C++ extension, that delivers the Window handle via API?

**Comment by Levitanus** [comment]
@dimentorium, here is [my branch](https://github.com/Levitanus/reapy/tree/feature/gui) with wrapped JS_ReaExtension. It does not support List* functions, but many of them looks like working correctly.
But, I have some troubles within handles managing. Especially in looking of the docker handler.
You can look how it's done in [SWS](https://github.com/reaper-oss/sws/blob/430602a9aa5fb8cbaa145ca9415efb57a3919013/sws_wnd.cpp).
I tried to copy the `DockWnd` class, but haven't succeed yet.

**Comment by DrewTNBD** [comment]
Hey crew. Just to add, if you use the getState and setState then you can use a Lua script within Reaper to set the state:
`reaper.SetExtState('section', 'key', 'value', false)`

And pick that up with Python like this:
```
import reapy as rp
from threading import Thread

class ReapyConnection:

    def __init__(self):
        self.project = rp.Project()
        self.poll = Thread(target=self.poll_For_Key, daemon=True).start()


    def poll_For_Key(self):
        while True:
            state = rp.get_ext_state('section', 'key')
            if len(state) > 0:
                print(state)
                rp.delete_ext_state('section', 'key')
            else:
                continue
        return
```

This has the benefit over TCP/IP in that your users don't need to have Reaper even be aware of Python.

@dimentorium 0 This method above works fine with Kivy as well btw. 

I suppose one thing to try (haven't done it yet) is wrapping the `poll_For_Key()` inside `with inside_reaper():` context manager to get the speed benefit.

**Comment by DrewTNBD** [comment]
Oh, and I'm reasonably disappointed in the Python support for Reaper. It seems to not be as loved as Lua. So I think this project is great, and anything I can do to help I will.

**Comment by RomeoDespres** [comment]
@DrewTNBD This kind of outer event loop checking for changes in ext state is indeed what I meant. It's probably less efficient than an outer HTTP server though.

Note that you can't wrap the entire outer event loop with `inside_reaper`, because if you do, REAPER will freeze. The REAPER UI freezes during the whole `inside_reaper` context. So I'm not sure you will have a huge performance benefit on this side.

> This has the benefit over TCP/IP in that your users don't need to have Reaper even be aware of Python.

Your users still have to have a reapy-enabled (and thus Python-enabled) REAPER don't they?

> Oh, and I'm reasonably disappointed in the Python support for Reaper.

Yup! In my opinion that is definitely something on which the REAPER team should work. Lua's certainly nice but Python is so widely used... I'm happy with reapy - but also disappointed somehow that we have to build it ourselves and always fight with the weird ReaScript interface.

**Comment by RomeoDespres** [comment]
> What about the idea building a basic framework for a Reapy based toolbox?
> So something with UI, event handling and communication to Reaper to make development of addons and plugins easier? From what I read it looks like we are building comparable stuff

This would be an amazing project!

**Comment by dimentorium** [comment]
As some open questions to think about.

Why do Reaper coders not like Python?
What would be needed in order to increase the user base?

If you want the Reaper team to invest more time into this, you need some really good use cases; at least I would think that. Currently as everyone is doing stuff in LUA and it works, they hav no need or benefit of investing more.

**Comment by Levitanus** [comment]
@dimentorium 
> Why do Reaper coders not like Python?

It's wrappers are buggy. And it requires nontrivial installation (which we will solve as we have a time for it)

> What would be needed in order to increase the user base?

make a tool that many needs, written on Python))

For me Python is just has more architectural tools, as well as huge codebase and std. Lua inside Reaper also is broken, but I'm confused, coders don't notice this.

**Comment by DrewTNBD** [comment]
> @DrewTNBD This kind of outer event loop checking for changes in ext state is indeed what I meant. It's probably less efficient than an outer HTTP server though.
> 
> Note that you can't wrap the entire outer event loop with `inside_reaper`, because if you do, REAPER will freeze. The REAPER UI freezes during the whole `inside_reaper` context. So I'm not sure you will have a huge performance benefit on this side.

Yep. I discovered this for myself yesterday when I tried it! lol!


> > This has the benefit over TCP/IP in that your users don't need to have Reaper even be aware of Python.
> 
> Your users still have to have a reapy-enabled (and thus Python-enabled) REAPER don't they?

Ah yes, doh! I'm being dumb on that point.

> > Oh, and I'm reasonably disappointed in the Python support for Reaper.
> 
> Yup! In my opinion that is definitely something on which the REAPER team should work. Lua's certainly nice but Python is so widely used... I'm happy with reapy - but also disappointed somehow that we have to build it ourselves and always fight with the weird ReaScript interface.

Fully agreed. I do think Reapy is a great piece of work and I'm very happy to have it.

**Comment by DrewTNBD** [comment]
> > What about the idea building a basic framework for a Reapy based toolbox?
> > So something with UI, event handling and communication to Reaper to make development of addons and plugins easier? From what I read it looks like we are building comparable stuff
> 
> This would be an amazing project!

I'd be happy to be involved in such a thing, as much as I'm clearly not on everyone else's level! lol. I'm just making mini tools for sample-pack development in-house with the BFD team (I worked for fxpansion/roli and now BFD is owned by inmusic)

That might complicate things, but if I can be involved then I will.

I see building a custom tool

**Comment by DrewTNBD** [comment]
(sorry for closing, fat fingers!)

**Comment by DrewTNBD** [comment]
> @dimentorium
> 
> > Why do Reaper coders not like Python?
> 
> It's wrappers are buggy. And it requires nontrivial installation (which we will solve as we have a time for it)
> 
> > What would be needed in order to increase the user base?
> 
> make a tool that many needs, written on Python))
> 
> For me Python is just has more architectural tools, as well as huge codebase and std. Lua inside Reaper also is broken, but I'm confused, coders don't notice this.

I know you can kinda do object orientated programming in Lua. But Python just feels more intuitive to me. Less messing around with end statements and relying solely on tables, and having access to more data-types like tuples and what not. Leveraging things like Numpy and SciPy to analyse audio time-series data and do FFT's and all that sort of thing.... it just boggles my mind a bit that Lua is the default language for Reascripting given how popular Python is and how easily extensible it is.

**Comment by Levitanus** [comment]
> I see building a custom tool to do a whole bunch of jobs as an attractive thing.

My dream. Where to keep track on your work? When I finish my — would like to join))

**Comment by dimentorium** [comment]
So I guess 3 topics:
1. Make installation easier => I can supply my first try for a portable python, maybe its a start
2. Make start easier => There are very few Python specific tutorials, that could be helpful
3. Killer Application => thats a tough one ;-) 

Maybe start some new threads for these

**Comment by DrewTNBD** [comment]
Kivy has on_start(), on_stop() and build() methods. I would imagine having Reaper launch and the local server connect up when you launch your Kivy app would be fairly trivial.

Then there is PyInstaller that can bundle a Python script(s) into a standalone executable.

Reaper supports portable installs, meaning if you could also distribute (licenses permitting) a copy of Reaper with it that is pre-setup to use the Python environment that PyInstaller bakes into it's resource files, then potentially you could have a one click solution for all this.

**Comment by RomeoDespres** [comment]
@dimentorium Thanks for summarizing the discussion. Indeed it probably make sens to talk about these points in separate discussions (such as #61).

I've also been thinking a lot lately about rewriting the docs entirely, because I find them pretty bad and especially lacking tutorials. Good to know you agree with that. I'll try to jot down a better structure for docs so that one can more easily inject stuff (e.g. tutos) into it.

**Comment by RomeoDespres** [comment]
> Reaper supports portable installs, meaning if you could also distribute (licenses permitting) a copy of Reaper with it that is pre-setup to use the Python environment that PyInstaller bakes into it's resource files, then potentially you could have a one click solution for all this.

This probably has good use cases, but I guess most users would want to install a plugin on top of their existing REAPER setup, rather than installing a whole new REAPER, don't you think?

**Comment by DrewTNBD** [comment]
> > Reaper supports portable installs, meaning if you could also distribute (licenses permitting) a copy of Reaper with it that is pre-setup to use the Python environment that PyInstaller bakes into it's resource files, then potentially you could have a one click solution for all this.
> 
> This probably has good use cases, but I guess most users would want to install a plugin on top of their existing REAPER setup, rather than installing a whole new REAPER, don't you think?

I'm reasonably sure you can still make a portable reaper setup look at the system VSTplugin folders.

**Comment by DrewTNBD** [comment]
Hey everyone, just in case you come across this thread in the future. My earlier "solution" of setting and getting Reaper extStates does work - but it only works for scripts run inside of Reaper.

I later tried to assign a function call to a button inside my kivy app, and it only worked half the time and for specific functions - getting the project and tracks worked, but getting items did not.

This seemed to be caused by my Reaper_Connection class that I wrote to poll the extStates. Because it was running in it's own thread, I ended up getting a bunch of utf-8 decode errors.

Don't know if that is a good explanation really. But I reverted over to the TCP solution provided by @dimentorium and that is working peachy.
